File: backend/engines/bayesian.py
import math

def _gauss_pdf(x: float, mu: float, sigma: float) -> float:
    if sigma <= 0:
        return 1.0
    return (1.0 / (sigma * math.sqrt(2 * math.pi))) * math.exp(-0.5 * ((x - mu) / sigma) ** 2)


def inferir_bayesiano(hechos: dict, variables: list[dict], clases: list[dict]) -> list[dict]:
    var_map = {v["nombre"]: v for v in variables}
    scores = []

    for cl in clases:
        score = float(cl.get("prior", 0.1))
        probs       = cl.get("probs", {})       # discretas: {"fiebre_si": 0.85}
        probs_gauss = cl.get("probs_gauss", {}) # gaussianas: {"temperatura": {"mu":38.9,"sigma":0.5}}

        for var_nombre, val in hechos.items():
            if val is None or val == "":
                continue
            var_def = var_map.get(var_nombre)
            if not var_def:
                continue

            if var_def["tipo"] == "numerico":
                g = probs_gauss.get(var_nombre)
                if g and g.get("sigma", 0) > 0:
                    p = _gauss_pdf(float(val), float(g["mu"]), float(g["sigma"]))
                    score *= max(p, 1e-10)
            else:
                key = f"{var_nombre}_{str(val).lower()}"
                p = float(probs.get(key, 0.5))  # Laplace smoothing: 0.5 si no definido
                score *= max(p, 1e-10)

        scores.append({"nombre": cl["nombre"], "score": score})

    total = sum(s["score"] for s in scores)
    result = [
        {
            "conclusion": s["nombre"],
            "certeza":    round(s["score"] / total, 4) if total > 0 else 0.0,
        }
        for s in scores
    ]
    return sorted(result, key=lambda x: -x["certeza"])

File: frontend/src/inference/bayesian.ts
// Motor bayesiano — Naive Bayes con PDF gaussiana para variables numéricas
// by Jose Rodas

function gaussPdf(x: number, mu: number, sigma: number): number {
  if (sigma <= 0) return 1.0;
  return (1 / (sigma * Math.sqrt(2 * Math.PI))) * Math.exp(-0.5 * ((x - mu) / sigma) ** 2);
}

export function inferirBayesiano(
  hechos: Record<string, any>,
  variables: any[],
  clases: any[],
): { conclusion: string; certeza: number }[] {
  const varMap = new Map(variables.map((v: any) => [v.nombre, v]));
  const scores: { nombre: string; score: number }[] = [];

  for (const cl of clases) {
    let score = parseFloat(cl.prior ?? 0.1);
    const probs      = cl.probs       ?? {};  // discretas: { "fiebre_si": 0.85 }
    const probsGauss = cl.probs_gauss ?? {};  // gaussianas: { "temp": { mu, sigma } }

    for (const [nombre, val] of Object.entries(hechos)) {
      if (val === undefined || val === '') continue;
      const varDef = varMap.get(nombre);
      if (!varDef) continue;

      if (varDef.tipo === 'numerico') {
        const g = probsGauss[nombre];
        if (g && parseFloat(g.sigma) > 0) {
          const p = gaussPdf(parseFloat(val as string), parseFloat(g.mu), parseFloat(g.sigma));
          score *= Math.max(p, 1e-10);
        }
      } else {
        const key = `${nombre}_${String(val).toLowerCase()}`;
        const p = parseFloat(probs[key] ?? 0.5);
        score *= Math.max(p, 1e-10);
      }
    }

    scores.push({ nombre: cl.nombre, score });
  }

  const total = scores.reduce((s, x) => s + x.score, 0);
  const result = scores.map(s => ({
    conclusion: s.nombre,
    certeza:    total > 0 ? Math.round((s.score / total) * 10000) / 10000 : 0,
  }));

  return result.sort((a, b) => b.certeza - a.certeza);
}
